{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tonotopy Analysis\n",
    "\n",
    "We would like to know if it's feasible to assign a cell to a tonotopycal region depending on its RNA expression.\n",
    "[Waldhaus et al., 2015](https://www.sciencedirect.com/science/article/pii/S2211124715004945?via%3Dihub) have shown that different cell types seem to have differential expression of some genes according to their position in the Apex to Base axis.\n",
    "Out first attempt at this determination is going to be:\n",
    "\n",
    "1. Choose a cell type\n",
    "2. Take the most DE genes across Apex vs Base\n",
    "3. Plot pair plots showing how the expression of these genes change across the Apex to Base axis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Remember to download the data from [Waldhaus et al., 2015](https://www.cell.com/cms/10.1016/j.celrep.2015.04.062/attachment/7d19c04f-89c9-4f8d-8c7d-670de495351a/mmc2.xls) to ```../data/Waldhaus_2015```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = Path(\"../data\")\n",
    "master_table_dir = DATA_DIR / \"Waldhaus_2015/mmc2.xls\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_excel(master_table_dir, skiprows=2)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now make a dictionary of all the genes of interest per region. This list includes genes that are differentially expressed between Middle region and Apex or Base, but not necessarily between Apex and Base."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_of_interest = {\n",
    "    \"GER\": [\n",
    "        \"Gapdh\",\n",
    "        \"Egfp\",\n",
    "        \"Acvr2b\",\n",
    "        \"Cacna1d\",\n",
    "        \"Cdh2\",\n",
    "        \"Cdk7\",\n",
    "        \"Clrn1\",\n",
    "        \"Dach1\",\n",
    "        \"Dkk3\",\n",
    "        \"Egfr\",\n",
    "        \"Fzd6\",\n",
    "        \"Gata3\",\n",
    "        \"Gli3\",\n",
    "        \"Gpr98\",\n",
    "        \"Hes5\",\n",
    "        \"Hmga2\",\n",
    "        \"Jag1\",\n",
    "        \"Lamb1\",\n",
    "        # \"Ncam\",\n",
    "        \"Ntf3\",\n",
    "        \"Sox2\"\n",
    "        ],\n",
    "    \"IHC\": [\n",
    "        \"Actb\",\n",
    "        \"Bmpr2\",\n",
    "        \"Calb1\",\n",
    "        \"Calb2\",\n",
    "        \"Calm1\",\n",
    "        \"Cdkn1a\",\n",
    "        \"Csnk1e\",\n",
    "        \"Dach1\",\n",
    "        \"Dkk3\",\n",
    "        \"Dner\",\n",
    "        \"Eya2\",\n",
    "        # \"Fgfr1\",\n",
    "        \"Fzd3\",\n",
    "        \"Gata3\",\n",
    "        \"Grb10\",\n",
    "        \"Isl1\",\n",
    "        \"Lfng\",\n",
    "        \"Ntf3\",\n",
    "        \"Otof\",\n",
    "        \"Pax2\",\n",
    "        \"Pdgfra\",\n",
    "        \"Ptprq\",\n",
    "        \"Pvalb\",\n",
    "        \"Sox2\",\n",
    "        \"Sox9\",\n",
    "        ],\n",
    "    \"IPH\": [\n",
    "        \"Cdk6\",\n",
    "        \"Fbxo2\",\n",
    "        \"Fgf10\",\n",
    "        \"Fos\",\n",
    "        \"Inhba\",\n",
    "        ],\n",
    "    \"IPC\": [\n",
    "        \"Actb\",\n",
    "        \"Calb2\",\n",
    "        \"Ccnd1\",\n",
    "        \"Cdh1\",\n",
    "        \"Fbxo2\",\n",
    "        \"Fgf10\",\n",
    "        \"Gjb1\",\n",
    "        \"Hmga2\",\n",
    "        \"Inhba\",\n",
    "        \"Itpr2\",\n",
    "        \"Kcnj10\",\n",
    "        \"Lgr5\",\n",
    "        \"Pax2\",\n",
    "        \"Pdgfra\",\n",
    "        ],\n",
    "    \"OPC\": [\n",
    "        \"Actb\",\n",
    "        \"Ai14-tdTomato\",\n",
    "        \"Acvr2b\",\n",
    "        \"Akt1\",\n",
    "        \"Aqp4\",\n",
    "        \"Bmp2\",\n",
    "        \"Calm1\",\n",
    "        \"Cdh2\",\n",
    "        \"Cdkn1a\",\n",
    "        \"Dach1\",\n",
    "        \"Dkk3\",\n",
    "        \"Efna1\",\n",
    "        \"Epha7\",\n",
    "        \"Espn\",\n",
    "        \"Etv4\",\n",
    "        \"Etv5\",\n",
    "        \"Fbxo2\",\n",
    "        \"Fgf10\",\n",
    "        # \"Fgfr1\",\n",
    "        # \"Fgfr3\",\n",
    "        \"Fos\",\n",
    "        \"Fst\",\n",
    "        \"Fzd3\",\n",
    "        \"Gdf10\",\n",
    "        \"Gjb1\",\n",
    "        \"Gli2\",\n",
    "        \"Gli3\",\n",
    "        \"Gpr98\",\n",
    "        \"Hes5\",\n",
    "        \"Hmga2\",\n",
    "        \"Inhba\",\n",
    "        \"Itpr2\",\n",
    "        \"Kcnj10\",\n",
    "        \"Lgr5\",\n",
    "        \"Myo6\",\n",
    "        \"Myo7a\",\n",
    "        \"Nes\",\n",
    "        \"Otol1\",\n",
    "        \"Pdgfra\",\n",
    "        \"Ptc1\",\n",
    "        \"Six4\",\n",
    "        \"Smo\",\n",
    "        \"Spry1\",\n",
    "        \"Tcf7l1\",\n",
    "        \"Tgfbr2\",\n",
    "        \"Vegfc\",\n",
    "        \"Wnt2b\",\n",
    "        \"Wnt7a\",\n",
    "        \"Wnt7b\",\n",
    "    ],\n",
    "    \"OHC\": [\n",
    "        \"Actb\",\n",
    "        \"Ai14-tdTomato\",\n",
    "        \"Acvr2b\",\n",
    "        \"Akt1\",\n",
    "        \"Bmpr2\",\n",
    "        \"Cacna1d\",\n",
    "        \"Calb1\",\n",
    "        \"Calb2\",\n",
    "        \"Calm1\",\n",
    "        \"Ccne1\",\n",
    "        \"Cdh1\",\n",
    "        \"Cdkn1a\",\n",
    "        \"Cdkn2d\",\n",
    "        \"Chrd\",\n",
    "        \"Clrn1\",\n",
    "        \"Csnk1e\",\n",
    "        \"Dach1\",\n",
    "        \"Dkk3\",\n",
    "        \"Dll1\",\n",
    "        \"Dnmt1\",\n",
    "        \"Dnmt3a\",\n",
    "        \"Etv5\",\n",
    "        \"Eya1\",\n",
    "        \"Eya2\",\n",
    "        \"Fbxo2\",\n",
    "        # \"Fgfr1\",\n",
    "        # \"Fgfr3\",\n",
    "        \"Fgfr4\",\n",
    "        \"Fos\",\n",
    "        \"Fst\",\n",
    "        \"Fzd3\",\n",
    "        \"Gata3\",\n",
    "        \"Grb10\",\n",
    "        \"Hey1\",\n",
    "        \"Hmga2\",\n",
    "        \"Isl1\",\n",
    "        \"Jag2\",\n",
    "        \"Lfng\",\n",
    "        \"Lgr5\",\n",
    "        \"Mcam\",\n",
    "        \"Myo15\",\n",
    "        \"Myo6\",\n",
    "        \"Myo7a\",\n",
    "        \"Ntf3\",\n",
    "        \"Ocm\",\n",
    "        \"Otof\",\n",
    "        \"Pax2\",\n",
    "        \"Pcdh15\",\n",
    "        \"Pcgf2\",\n",
    "        \"Pdgfra\",\n",
    "        \"Prox1\",\n",
    "        \"Ptc1\",\n",
    "        \"Ptprq\",\n",
    "        \"Six4\",\n",
    "        \"Slc26a5\",\n",
    "        \"Sox2\",\n",
    "        \"Sox9\",\n",
    "        \"Spna2\",\n",
    "        \"Spry2\",\n",
    "        \"Vegfa\",\n",
    "        \"Wnt2b\",\n",
    "        \"Wnt7a\",\n",
    "    ],\n",
    "    \"DC12\": [\n",
    "        \"Cdh1\",\n",
    "        \"Csnk1e\",\n",
    "        \"Egfr\",\n",
    "        \"Gjb1\",\n",
    "        \"Hmga2\",\n",
    "        \"Inhba\",\n",
    "        \"Isl1\",\n",
    "        \"Itpr2\",\n",
    "        \"Kcnj10\",\n",
    "        \"Lamb1\",\n",
    "        \"Map2k1\",\n",
    "        \"Mcam\",\n",
    "        \"Myo6\",\n",
    "        \"Otor\",\n",
    "        \"Tcf7l1\",\n",
    "    ],\n",
    "    \"DC3\": [\n",
    "        \"Cdh1\",\n",
    "        \"Cdk6\",\n",
    "        \"Fgf10\",\n",
    "        \"Fgfr4\",\n",
    "        \"Hmga2\",\n",
    "        \"Inhba\",\n",
    "        \"Kcnj10\",\n",
    "        \"Myo7a\",\n",
    "        \"Tcf7l1\",\n",
    "    ]\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have the genes of interest per cell type, we can make pairplots and try to see if gene expression correlates with tonotopy in some clear way.\n",
    "\n",
    "Commented out genes were not found in the master table. Careful because it takes a while to generate tge following plots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for cell_type, cell_type_de_genes in genes_of_interest.items():\n",
    "    print(cell_type)\n",
    "    cols = [\"Anatomical Origin\"] + cell_type_de_genes\n",
    "    selected_df = df.query(f\"Subpopulation == '{cell_type}'\")[cols]\n",
    "\n",
    "    sns.pairplot(data=selected_df, hue=\"Anatomical Origin\", plot_kws={\"rasterized\":True}, diag_kws={\"rasterized\":True})\n",
    "    plt.title(cell_type)\n",
    "    plt.savefig(DATA_DIR.parent.parent / f\"results/waldhaus_2015_analysis/{cell_type}.png\", format=\"png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We should save a dictionary showing which genes show a higher expression in each cell type.\n",
    "This dictionary we will use later to name and corroborate that each group assigned in our dataset corresponds to one of these groups and that the same pattern of expression is repeated in both groups.\n",
    "We will have to skip the cells assigned to Middle as then we can't assign anything here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "waldhaus_high_expression_dict = dict()\n",
    "\n",
    "for cell_type, cell_type_de_genes in genes_of_interest.items():\n",
    "    print(cell_type)\n",
    "    cols = [\"Anatomical Origin\"] + cell_type_de_genes\n",
    "    selected_df = df.dropna(subset=\"Anatomical Origin\").query(f\"Subpopulation == '{cell_type}'\")[cols]\n",
    "    selected_df = selected_df[~selected_df[\"Anatomical Origin\"].str.contains(\"Middle\")]\n",
    "    \n",
    "    waldhaus_high_expression_dict[cell_type] = selected_df.groupby(\"Anatomical Origin\").mean().idxmax()\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load our dataset\n",
    "\n",
    "We can load expression of landmark genes in our particular cell types and look into separating them with a Gaussian Mixture Model.\n",
    "This will yield two separate populations of cells.\n",
    "We can then assess which genes have higher expression in each subpopulation and then assign Apex or Base to each class.\n",
    "Finally, we can obtain the probability of belonging to each of these classes and use this to classify cells.\n",
    "\n",
    "After looking into gene expression in our datasets, we noticed that gene expression of OHC depended on collection date so we kept only datasets collected in P1.\n",
    "We also noticed that some OHC cells showed a very low expression of Calm1 (<4.5) which we also filtered out. Something similar was happenning with IHC which expressed low Pvalb (<2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "datasets = dict()\n",
    "sheets_dict = pd.read_excel(DATA_DIR / \"expression_tonogenes_trex.xlsx\", sheet_name=None)\n",
    "\n",
    "for this_cell_type, this_df in sheets_dict.items():\n",
    "    print(this_cell_type)\n",
    "    \n",
    "    this_df[\"dataset\"] = this_df[this_df.columns[0]].apply(lambda x: int(x.split(\"_\")[-1]))\n",
    "    this_df = this_df[this_df.dataset.isin([3, 4, 5])]\n",
    "\n",
    "    if this_cell_type == \"OHC\":\n",
    "        this_df = this_df.query(\"Calm1 > 4.5\")\n",
    "    if this_cell_type == \"IHC\":\n",
    "        this_df = this_df.query(\"Pvalb > 2\")\n",
    "    \n",
    "    datasets[this_cell_type] = this_df\n",
    "datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cluster cells into Apex and Base\n",
    "\n",
    "The goal is to use Gaussian Mixture Models to segregate between Apex and Base for each cell type.\n",
    "We will use the gene expression found in Waldhaus to initiate the GMMs near the expected expression pattern for each gene from each region for each cell type.\n",
    "Briefly, for each cell type, we will look into each gene and find the minimum and maximum expression level in our dataset.\n",
    "We will initialize the GMM with the means so that one group should start near what would be expected for an Apex cell expressing the maximal values and then the analogous for a Base cell.\n",
    "Once GMM converges for each cell type, we will assess that the expression distribution of each cell type matches the Waldhaus paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.mixture import GaussianMixture"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We made a list of all the genes that have a clear differential expression in Waldhaus's dataset and also tagged in which region the highest expression was happenning."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_expression_gene_dict = {\n",
    "    \"GER\": {\n",
    "        \"Dkk3\": \"Base\",\n",
    "        \"Jag1\": \"Apex\"\n",
    "    },\n",
    "    \"IPH\": {\n",
    "        \"Fbxo2\": \"Apex\",\n",
    "        \"Fos\": \"Base\",\n",
    "    },\n",
    "    \"IHC\": {\n",
    "        \"Actb\": \"Apex\",\n",
    "        \"Calb2\": \"Base\",\n",
    "        \"Dach1\": \"Apex\",\n",
    "        \"Gata3\": \"Base\",\n",
    "        \"Fzd3\": \"Apex\",\n",
    "        \"Isl1\": \"Apex\",\n",
    "        \"Ntf3\": \"Apex\",\n",
    "        \"Otof\": \"Base\",\n",
    "        \"Pax2\": \"Apex\",\n",
    "        \"Ptprq\": \"Base\",\n",
    "        \"Pvalb\": \"Base\",\n",
    "        \"Sox2\": \"Apex\",\n",
    "    },\n",
    "    \"IPC\": {\n",
    "        \"Ccnd1\": \"Apex\",\n",
    "        \"Hmga2\": \"Apex\",\n",
    "        \"Inhba\": \"Base\",\n",
    "        \"Kcnj10\": \"Base\"\n",
    "    },\n",
    "    \"OPC\": {\n",
    "        \"Actb\": \"Apex\",\n",
    "        \"Fbxo2\": \"Apex\",\n",
    "        \"Hes5\": \"Apex\",\n",
    "        \"Itpr2\": \"Base\",\n",
    "        \"Tcf7l1\": \"Base\",\n",
    "    },\n",
    "    \"OHC\": {\n",
    "        \"Actb\": \"Apex\",\n",
    "        \"Cacna1d\": \"Base\",\n",
    "        \"Calb1\": \"Base\",\n",
    "        \"Calb2\": \"Base\",\n",
    "        \"Clm1\": \"Base\",\n",
    "        \"Eya1\": \"Apex\",\n",
    "        \"Myo6\": \"Base\",\n",
    "        \"Otof\": \"Base\",\n",
    "        \"Pcdh15\": \"Apex\",\n",
    "        \"Ptprq\": \"Base\"\n",
    "    },\n",
    "    \"DC12\": {\n",
    "        \"Inhba\": \"Base\",\n",
    "        \"Map2k1\": \"Apex\"\n",
    "    },\n",
    "    \"DC3\": {\n",
    "        \"Fg10\": \"Apex\",\n",
    "        \"Hmga2\": \"Apex\",\n",
    "        \"Inhba\": \"Base\",\n",
    "        \"Tcf7l1\": \"Apex\",\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gmms = dict()\n",
    "\n",
    "for this_cell_type, this_df in datasets.items():\n",
    "    print(f\"Processing cell type: {this_cell_type}\")\n",
    "    \n",
    "    this_gene_list = this_df.columns.drop([\"cellID\", \"dataset\"]).values\n",
    "    \n",
    "    if this_cell_type == \"GER\":\n",
    "        this_gene_list = [\"Dach1\", \"Dkk3\"]\n",
    "    \n",
    "    if this_cell_type == \"IHC\":\n",
    "        this_gene_list = [\"Actb\", \"Calb1\", \"Isl1\", \"Otof\", \"Pvalb\"]\n",
    "    \n",
    "    if this_cell_type == \"IPH\":\n",
    "        # Only has the 4 dark green genes\n",
    "        pass\n",
    "    \n",
    "    if this_cell_type == \"IPC\":\n",
    "        this_gene_list = [\"Fbxo2\", \"Hmga2\", \"Inhba\",\"Itpr2\", \"Kcnj10\"]\n",
    "    \n",
    "    if this_cell_type == \"OPC\":\n",
    "        this_gene_list = [\"Fzd3\", \"Inhba\", \"Itpr2\", \"Tcf7l1\"]\n",
    "    \n",
    "    if this_cell_type == \"OHC\":\n",
    "        this_gene_list = this_df.columns.drop([\"cellID\", \"dataset\", \"Fbxo2\", \"Gata3\"]).values\n",
    "\n",
    "    print(\"We will use the following genes to fit GMMs\")\n",
    "    print(this_gene_list)\n",
    "    \n",
    "    gene_min_max = this_df[this_gene_list].agg([min, max]).values\n",
    "\n",
    "    for n, should_swap in enumerate([False  if region ==\"Apex\" else True for region in waldhaus_high_expression_dict[this_cell_type][this_gene_list]]):\n",
    "        if should_swap:\n",
    "            gene_min_max[:, n] = np.flip(gene_min_max[:, n])\n",
    "    \n",
    "    gmms[this_cell_type] = GaussianMixture(n_components=2, means_init=gene_min_max, max_iter=1000, random_state=42)\n",
    "    gmms[this_cell_type].fit(this_df[this_gene_list])\n",
    "    if not gmms[this_cell_type].converged_:\n",
    "        print(\"GMM has not converged\")\n",
    "    \n",
    "    this_df[\"predicted\"] = gmms[this_cell_type].predict(this_df.loc[:, this_gene_list])\n",
    "\n",
    "    sns.pairplot(this_df, hue=\"predicted\")\n",
    "    plt.suptitle(this_cell_type)\n",
    "    plt.savefig(DATA_DIR.parent / f\"results/gmm_initialized_{this_cell_type}.png\", format=\"png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once we have GMMs trained for predicting to which region the different cell types belong to, we have to assess whether the gene expression of these predictions match the gene pattern observed in Wladhaus' datasets.\n",
    "\n",
    "Let's make a function to parse the group number into the most likely region.\n",
    "We can then check which genes seem to not follow the pattern and assess if these clearly had a differential gene expression or if they could be mislabeled."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_group_dict(assignment_counts):\n",
    "    most_common = assignment_counts.most_common()[0][0]\n",
    "    \n",
    "    if not most_common[0] in [0, 1]:\n",
    "        raise ValueError(\"Number not in 0 or 1\")\n",
    "    if not most_common[1] in [\"Base\", \"Apex\"]:\n",
    "        raise ValueError(\"Region not in Base or Apex\")\n",
    "    \n",
    "    other_number = 0 if most_common[0] == 1 else 1\n",
    "    other_region = \"Apex\" if most_common[1] == \"Base\" else \"Base\"\n",
    "\n",
    "    group_dict = {\n",
    "        most_common[0]: most_common[1],\n",
    "        other_number: other_region\n",
    "    }\n",
    "\n",
    "    return group_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "detected_high_expression_dict = dict()\n",
    "\n",
    "for this_cell_type, this_gmm in gmms.items():\n",
    "    print(this_cell_type)\n",
    "    detected_high_expression_dict[this_cell_type] = pd.Series(np.argmax(this_gmm.means_, axis=0), index=this_gmm.feature_names_in_)\n",
    "\n",
    "detected_high_expression_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "group_dicts = dict()\n",
    "\n",
    "for this_cell_type, this_gmm in gmms.items():\n",
    "    print(this_cell_type)\n",
    "\n",
    "    this_gene_list = detected_high_expression_dict[this_cell_type].index\n",
    "    assignments = [(detected, waldhaus) for detected, waldhaus in zip(detected_high_expression_dict[this_cell_type], waldhaus_high_expression_dict[this_cell_type][this_gene_list])]\n",
    "\n",
    "    assignment_counts = Counter(assignments)\n",
    "    group_dicts[this_cell_type] = get_group_dict(assignment_counts)\n",
    "\n",
    "    landmarked_genes = list(high_expression_gene_dict[this_cell_type].keys())\n",
    "    this_high_df = detected_high_expression_dict[this_cell_type].rename(\"detected\").to_frame().join(waldhaus_high_expression_dict[this_cell_type][this_gene_list].rename(\"waldhaus\"))\n",
    "    this_high_df[\"detected\"] = this_high_df[\"detected\"].apply(lambda x: group_dicts[this_cell_type][x])\n",
    "    this_high_df[\"landmarked_gene\"] = this_high_df.index.map(lambda x: x in landmarked_genes)\n",
    "\n",
    "    print(this_high_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From comparing these, not all genes seem to follow the same pattern in a straightforward way.\n",
    "We should also take into account that not all genes show an expression different enough for us to be confident that one is higher than the other, for this reason we have taggedd some genes that should be clear landmark genes.\n",
    "\n",
    "By looking into this list, GER is the only cell type which we cannot be sure about its assignment.\n",
    "For the other cell types, there are very few (2) mismatches in expression, but these are in genes that are not really differentially expressed.\n",
    "\n",
    "Finally, let's apply the trained GMMs to predict region and also state how certain we are of belonging to such region."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.ExcelWriter(DATA_DIR.parent / f\"results/tonotopy_initialized_prediction_2.xlsx\") as writer:\n",
    "    for this_cell_type, this_df in datasets.items():\n",
    "        print(this_cell_type)\n",
    "\n",
    "        this_gene_list = detected_high_expression_dict[this_cell_type].index\n",
    "\n",
    "        this_gmm = gmms[this_cell_type]\n",
    "        this_group_dict = group_dicts[this_cell_type]\n",
    "        \n",
    "        this_df.loc[:, \"predicted\"] = [this_group_dict[x] for x in gmms[this_cell_type].predict(this_df.loc[:, this_gene_list])]\n",
    "        this_df[[f\"proba_{this_group_dict[0]}\", f\"proba_{this_group_dict[1]}\"]] = gmms[this_cell_type].predict_proba(this_df.loc[:, this_gene_list])\n",
    "\n",
    "        this_df.to_excel(excel_writer=writer, sheet_name=this_cell_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plots\n",
    "\n",
    "Having a updataed dataframes with all the predictions and probabilities, we can plot:\n",
    "\n",
    "1. A pair plot separated by regions, so that we can vissually assess expression of each of the genes.\n",
    "\n",
    "2. A pair plot where probability is color coded, so that we can see how fast probability changes across expression of one of the genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for this_cell_type, this_df in datasets.items():\n",
    "    print(this_cell_type)\n",
    "    sns.pairplot(this_df, hue=\"predicted\")\n",
    "    plt.suptitle(this_cell_type)\n",
    "    plt.savefig(DATA_DIR.parent.parent / f\"results/waldhaus_2015_analysis/initialized_prediction_{this_cell_type}.png\", format=\"png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for this_cell_type, this_df in datasets.items():\n",
    "    print(this_cell_type)\n",
    "    sns.pairplot(this_df, hue=\"proba_Apex\")\n",
    "    plt.suptitle(this_cell_type)\n",
    "    plt.savefig(DATA_DIR.parent.parent / f\"results/waldhaus_2015_analysis/proba_initialized_prediction_{this_cell_type}.png\", format=\"png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The question arose, if we could filter out cells which we were unsure about the region and get a very segregated gene expression distribution for each cell type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "probability_threshold = 0.9999\n",
    "\n",
    "for this_cell_type, this_df in datasets.items():\n",
    "    print(this_cell_type)\n",
    "    sns.pairplot(this_df.query(f\"proba_Apex > {probability_threshold} or proba_Base > {probability_threshold}\"),\n",
    "                 hue=\"proba_Apex\",\n",
    "                 plot_kws={\"alpha\": 0.6})\n",
    "    plt.suptitle(this_cell_type)\n",
    "    plt.savefig(DATA_DIR.parent.parent / f\"results/waldhaus_2015_analysis/certain_proba_initialized_prediction_{this_cell_type}.png\", format=\"png\")\n",
    "    plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Validation\n",
    "\n",
    "To validate our pipeline, we can train GMMs on Waldhaus' datasets, predict to which region each cell belongs to and assess our accurracy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "waldhaus_ggms = dict()\n",
    "waldhaus_results = dict()\n",
    "\n",
    "for this_cell_type, other_gmm in gmms.items():\n",
    "    print(this_cell_type)\n",
    "    this_gene_list = other_gmm.feature_names_in_\n",
    "\n",
    "    cols = [\"Anatomical Origin\"] + this_gene_list.tolist()\n",
    "    this_df = df.query(f\"Subpopulation == '{this_cell_type}'\")[cols].copy()\n",
    "    this_df = this_df.dropna(subset=\"Anatomical Origin\")\n",
    "    this_df = this_df.query(\"`Anatomical Origin` != 'Middle'\")\n",
    "    \n",
    "    if this_cell_type == \"OHC\":\n",
    "        this_df.query(\"Cacna1d > 2\")\n",
    "    \n",
    "    \n",
    "    gene_min_max = this_df[this_gene_list].agg([min, max]).values\n",
    "\n",
    "    for n, should_swap in enumerate([False  if region ==\"Apex\" else True for region in waldhaus_high_expression_dict[this_cell_type][this_gene_list]]):\n",
    "        if should_swap:\n",
    "            gene_min_max[:, n] = np.flip(gene_min_max[:, n])\n",
    "    \n",
    "    waldhaus_ggms[this_cell_type] = GaussianMixture(n_components=2, means_init=gene_min_max, max_iter=1000, random_state=42)\n",
    "    waldhaus_ggms[this_cell_type].fit(this_df[this_gene_list])\n",
    "    if not waldhaus_ggms[this_cell_type].converged_:\n",
    "        print(\"GMM has not converged\")\n",
    "    \n",
    "    this_df[\"predicted\"] = waldhaus_ggms[this_cell_type].predict(this_df.loc[:, this_gene_list])\n",
    "    waldhaus_results[this_cell_type] = this_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "waldhaus_group_dicts = dict()\n",
    "\n",
    "for this_cell_type, this_gmm in waldhaus_ggms.items():\n",
    "    print(this_cell_type)\n",
    "\n",
    "    high_expression_list = pd.Series(np.argmax(this_gmm.means_, axis=0), index=this_gmm.feature_names_in_)\n",
    "\n",
    "    this_gene_list = this_gmm.feature_names_in_\n",
    "    assignments = [(detected, waldhaus) for detected, waldhaus in zip(high_expression_list, waldhaus_high_expression_dict[this_cell_type][this_gene_list])]\n",
    "\n",
    "    assignment_counts = Counter(assignments)\n",
    "    waldhaus_group_dicts[this_cell_type] = get_group_dict(assignment_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "for this_cell_type, this_result in waldhaus_results.items():\n",
    "    print(this_cell_type)\n",
    "\n",
    "    this_group_dict = group_dicts[this_cell_type]\n",
    "    this_gene_list = waldhaus_ggms[this_cell_type].feature_names_in_\n",
    "\n",
    "    this_result.loc[:, \"predicted\"] = [this_group_dict[x] for x in waldhaus_ggms[this_cell_type].predict(this_result.loc[:, this_gene_list])]\n",
    "    print(pd.crosstab(this_result[\"Anatomical Origin\"], this_result[\"predicted\"]))\n",
    "    print(\"=========================================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for this_cell_type, this_df in waldhaus_results.items():\n",
    "    print(this_cell_type)\n",
    "    sns.pairplot(this_df, hue=\"predicted\")\n",
    "    plt.suptitle(this_cell_type)\n",
    "    plt.savefig(DATA_DIR.parent.parent / f\"results/waldhaus_2015_analysis/algorithm_validation_{this_cell_type}.png\", format=\"png\")\n",
    "    plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
